Social: Landing page pricing tweaks.

spsiddarthan opened the pull request:
This PR enables you to select multiple plans of Jetpack Social.  The variant selection for checking out the advanced plan isn't working currently on calypso blue correctly. It doesn't show the monthly variant when you go to the checkout page
#### Proposed Changes

* The changes would have been a lot simpler if we decided to do the same thing that backup does, but since we decided to not show the radio buttons, the product-lightbox changes needed to be a little bit more social specific, as opposed to being generic. The changes are mainly around the multiplechoice question component and adding some additional props to the PaymentPlan component.
* I have introduced the `jetpack-social/advanced-plan` feature flag, so the changes are picked up only on local dev environment.

How Social looks:

<img width="1226" alt="Screenshot 2023-01-13 at 18 49 09" src="https://user-images.githubusercontent.com/6594561/212329252-94f6c789-f789-4a33-bf19-5a2c8c907396.png">

How Backup looks:

<img width="1249" alt="Screenshot 2023-01-13 at 18 50 03" src="https://user-images.githubusercontent.com/6594561/212329382-98f7d1f8-3748-43e7-af4a-550b1b2f3ffe.png">

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Checkout to this branch and do `yarn start-jetpack-cloud` and go to http://jetpack.cloud.localhost:3000/pricing?view=products
*  Clicking on Social's GET button and more about Social button should open the modal.
* After opening the modal, select the plan of your choice and finish checkout.

To test plan's variant selections b/w the monthly and yearly plans for the advanced plan
*  Do `yarn start` to boot calypso
* Go to http://calypso.localhost:3000/checkout/jetpack/jetpack_social_advanced_yearly to make sure the monthly variant is shown as well. 
<img width="553" alt="Screenshot 2023-01-13 at 18 56 16" src="https://user-images.githubusercontent.com/6594561/212330552-5d27894e-2365-45a1-a9c0-d5ca80e4a9bb.png">

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

spsiddarthan committed:
Added lightbox options for Jetpack Social plans.

spsiddarthan committed:
Added variant selection for Jetpack Social basic and advanced plans.

spsiddarthan committed:
Fixed the issue with variant selection b/w yearly and monthly plans.

spsiddarthan committed:
Added features and benefits.

spsiddarthan committed:
Added config key to hide the feature till the time it's ready for use.

spsiddarthan committed:
Allow for variant selection.

spsiddarthan committed:
Use old flow when the flag is false.

spsiddarthan committed:
Clicking on GET should open the dialog.

spsiddarthan committed:
Show social plans one below the other.

spsiddarthan committed:
Fix plan names.

spsiddarthan committed:
Get unit tests to pass.

spsiddarthan commented:
Some unit tests are failing, I need to figure out what they're but will appreciate a review, so I can address them as well as I figure out the unit tests.

danielpost commented on client/my-sites/plans/jetpack-plans/product-lightbox/constants.ts:
```diff
@@ -19,11 +21,19 @@ export const PRODUCT_OPTIONS: Record< string, TranslateResult > = {
 		comment: 'GB here is the abbreviation for Gigabyte, and TB is abbreviation for Terabyte',
 	} ),
+	[ PRODUCT_JETPACK_SOCIAL_BASIC ]: translate( 'Social Basic', {
+		comment: 'Get upto 1000 shares',
```
"upto" -> "up to".

danielpost commented on client/my-sites/plans/jetpack-plans/product-lightbox/index.tsx:
```diff
@@ -71,6 +78,24 @@ const ProductLightbox: React.FC< Props > = ( {
+			);
+		},
+		[ onChangeProduct, dispatch, siteId ]
+	);
```
Is it necessary to add this function separately for Social products? It seems identical to the `onChangeOption` function that's declared right above it.

spsiddarthan replied:
You're right, we don't need this function

danielpost replied:
The typecheck `: string` got removed, is that intentional?

danielpost commented on client/my-sites/plans/jetpack-plans/product-lightbox/index.tsx:
```diff
@@ -108,7 +133,14 @@ const ProductLightbox: React.FC< Props > = ( {
+
+	const shouldShowOptions = isSocialProduct
+		? variantOptions.length > 1 && config.isEnabled( 'jetpack-social/advanced-plan' )
+		: variantOptions.length > 1;
```
I'm guessing this is temporary because the Advanced plan isn't available yet, right? If so, maybe we can add a comment here and add an Asana task to remove it once the plan is launched.

danielpost commented on client/my-sites/plans/jetpack-plans/product-lightbox/index.tsx:
```diff
@@ -176,7 +208,7 @@ const ProductLightbox: React.FC< Props > = ( {
 					<div className="product-lightbox__variants">
 						<div className="product-lightbox__variants-content">
-							{ shouldShowOptions && (
+							{ shouldShowOptions && ! isSocialProduct && (
```
Doesn't this extra `! isSocialProduct` check render the changes to `shouldShowOptions` made earlier obsolete?

spsiddarthan replied:
I changed the variable name here and it looks like this now, let me know if that makes sense. 

```
{ shouldShowOptions && ! showVariantsWithoutRadioButtons && (
								<div className="product-lightbox__variants-options">
									<MultipleChoiceQuestion
										question={ PRODUCT_OPTIONS_HEADER[ product?.productSlug ] }
										answers={ variantOptions }
										selectedAnswerId={ product?.productSlug }
										onAnswerChange={ onChangeOption }
										shouldShuffleAnswers={ false }
									/>
								</div>
							) }
							{ shouldShowOptions && showVariantsWithoutRadioButtons ? ( .....
```

danielpost commented on client/my-sites/plans/jetpack-plans/product-lightbox/index.tsx:
```diff
@@ -187,11 +219,52 @@ const ProductLightbox: React.FC< Props > = ( {
-								siteId={ siteId }
-								product={ product }
-							/>
+							{ isSocialProduct ? (
```
I'm wondering if we want to turn this into something more generic, since we might add more products with multiple plan options in the future.

What do you think about adding some kind of check that checks if the product has multiple plan options, rather than checking if it's a Social product?

spsiddarthan replied:
Thanks, good point, the whole change is around how we currently display products with multiple plans. 

The existing code does support displaying multiple products, backup works like that, for instance. 

But the way we display the multiple products for Social and backup is different, for the latter, we use a radio button. For Social, we are displaying them, one below the other. So, maybe we have two types of modals - one for the products where we should have the radio button and the other for products like social. I will make this have a more generic name.

danielpost commented on client/my-sites/plans/jetpack-plans/product-lightbox/payment-plan.tsx:
```diff
@@ -13,11 +13,15 @@ type PaymentPlanProps = {
 	siteId,
 	product,
+	hideTitle = false,
+	showBorder = false,
```
Maybe we can describe the state here rather than the UI—something like `isActive`?

danielpost commented on client/my-sites/plans/jetpack-plans/product-lightbox/payment-plan.tsx:
```diff
@@ -61,9 +65,15 @@ const PaymentPlan: React.FC< PaymentPlanProps > = ( {
+					{ ! hideTitle && <p>{ translate( 'Payment plan:' ) }</p> }
 
-					<div className="product-lightbox__variants-plan-card">
+					<div
+						className={
+							showBorder
+								? 'product-lightbox__variants-plan-card'
+								: 'product-lightbox__variants-plan-card noborder'
+						}
+					>
```
You could use the `classNames` helper here instead (untested):
```js
<div className={ classNames(
  'product-lightbox__variants-plan-card',
  ! showBorder && 'product-lightbox__variants-plan-card noborder'
) }>
```

I think here it's also better to describe the state rather than just the UI change, so I'd recommend something like this:
```js
<div className={ classNames(
  'product-lightbox__variants-plan-card',
  isActive && 'product-lightbox__variants-plan-card--active'
) }>
```

(and then change the class in the CSS file as well of course)

danielpost commented on client/my-sites/plans/jetpack-plans/product-store/items-list/all-items.tsx:
```diff
@@ -95,14 +96,24 @@ export const AllItems: React.FC< AllItemsProps > = ( {
 							<SimpleItemCard
 								ctaAsPrimary={ ctaAsPrimary }
-								ctaHref={ getCheckoutURL( item ) }
+								ctaHref={
+									item.productSlug === PRODUCT_JETPACK_SOCIAL_BASIC &&
+									config.isEnabled( 'jetpack-social/advanced-plan' )
+										? '#'
+										: getCheckoutURL( item )
+								}
```
I think it would be good to move this logic to a separate `const` declaration above the return statement, with a comment explaining what it's doing.

danielpost commented on client/my-sites/plans/jetpack-plans/product-store/items-list/all-items.tsx:
```diff
@@ -95,14 +96,24 @@ export const AllItems: React.FC< AllItemsProps > = ( {
 								isCtaDisabled={ isCtaDisabled }
 								isCtaExternal={ isExternal }
-								onClickCta={ getOnClickPurchase( item ) }
+								onClickCta={
+									item.productSlug === PRODUCT_JETPACK_SOCIAL_BASIC &&
+									config.isEnabled( 'jetpack-social/advanced-plan' )
+										? onClickMoreInfoFactory( item )
+										: getOnClickPurchase( item )
+								}
```
Same here as last comment :)

danielpost commented on packages/calypso-products/src/constants/jetpack.ts:
```diff
@@ -102,12 +104,18 @@ export const JETPACK_SCAN_PRODUCTS = <const>[
+	PRODUCT_JETPACK_SOCIAL_ADVANCED,
+	PRODUCT_JETPACK_SOCIAL_ADVANCED_MONTHLY,
+];
+
```
`JETPACK_SOCIAL_PRODUCTS` contains the Basic products and `JETPACK_SOCIAL_ADVANCED_PLANS` contains the Advanced products. It would probably be good to make these names consistent.

danielpost replied:
Should we also change `JETPACK_SOCIAL_PRODUCTS` to `JETPACK_SOCIAL_BASIC_PRODUCTS`?

spsiddarthan replied:
Sure thing, it is being used in many files though, and the last time we tried tweaking that, we broke the pricing page. Due to that reason, I wonder if we can do that renaming in a different PR as a follow-up and test that separately, to reduce risk.

danielpost commented on packages/calypso-products/src/constants/jetpack.ts:
```diff
@@ -144,6 +152,7 @@ export const JETPACK_PRODUCTS_LIST = <const>[
 	...JETPACK_VIDEOPRESS_PRODUCTS,
 	...JETPACK_SOCIAL_PRODUCTS,
 	...JETPACK_BACKUP_ADDON_PRODUCTS,
+	...JETPACK_SOCIAL_ADVANCED_PLANS,
```
Very nitpicky, but let's put the lines with the Social products right below each other.

danielpost reviewed (commented):
Nice work! I've added some initial thoughts from reading the code, will give it a functional test next.

spsiddarthan committed:
Fix failing unit test.

spsiddarthan committed:
Address review comments.

danielpost commented:
Pressing GET on the pricing page takes me directly to the checkout page (using the live link for testing), rather than opening the modal. Is that a bug?

spsiddarthan committed:
More review comments addressed.

spsiddarthan commented:
> Pressing GET on the pricing page takes me directly to the checkout page (using the live link for testing), rather than opening the modal. Is that a bug?

It's because of the env variable `jetpack-social/advanced-plan` which is enabled only for the development mode, meaning it will work only local. I will find out how to enable it for the live url sites, it should be a matter of changing one of the config files.

spsiddarthan committed:
Make variable name generic.

spsiddarthan committed:
Use class name helper.

spsiddarthan committed:
enable the feature flag on cloud horizon.

danielpost reviewed (commented):
Nice, just some tiny things!

I noticed this part is empty: 
![Screenshot 2023-01-17 at 14 53 10@2x](https://user-images.githubusercontent.com/1713699/212916651-9663182a-c724-441f-9619-7c174febf6b2.png)

Is that something we need to add in this PR as well?

spsiddarthan committed:
Code review comments and change name to Advanced Beta.

spsiddarthan commented:
@danielpost, I have fixed the empty "Great for" and changed the name to "Advanced Plan Beta", per Kristina's recommendation. We might go tweak the name, but I think we can always do that later.

danielpost reviewed (approved):
Looks good to me, nice work! Maybe @pablinos or @gmjuhasz can take a look as well just to be safe.

pablinos commented on client/my-sites/plans/jetpack-plans/product-lightbox/index.tsx:
```diff
@@ -57,7 +64,7 @@ const ProductLightbox: React.FC< Props > = ( {
 
 	const onChangeOption = useCallback(
-		( productSlug: string ) => {
+		( productSlug ) => {
```
I think @danielpost asked this, but is `productSlug` no longer a string? If so, can we update the type rather than removing it?

pablinos commented on client/my-sites/plans/jetpack-plans/product-lightbox/index.tsx:
```diff
@@ -108,7 +118,15 @@ const ProductLightbox: React.FC< Props > = ( {
 	}, [ product.productSlug ] );
 
-	const shouldShowOptions = variantOptions.length > 1;
+	const showVariantsWithoutRadioButtons = [
```
This is a bit of a nitpick (and I'm struggling to think of a better option) but this name and the `showPlansOneBelowTheOther` seem a bit too tied to the design.

Maybe it's something for a follow-up but all the conditional checks might mean that we could break this into multiple components and have two styles for the product variants.

manzoorwanijk replied:
I agree with the suggestion.

spsiddarthan replied:
I have made it a component, but the variable name I am not convinced, but I can definitely think of a better name and update it in the future @pablinos.

pablinos reviewed (commented):
I've only given this a quick look so far. I wonder if it's worth getting a review from Origin if we can?

manzoorwanijk commented on client/my-sites/plans/jetpack-plans/product-lightbox/index.tsx:
```diff
@@ -187,11 +205,48 @@ const ProductLightbox: React.FC< Props > = ( {
-								siteId={ siteId }
-								product={ product }
-							/>
+							{ shouldShowOptions && showVariantsWithoutRadioButtons ? (
+								<div>
+									<FormFieldset className="multiple-choice-question">
+										<FormLegend className="multiple-choice-question-title">
+											{ PRODUCT_OPTIONS_HEADER[ product?.productSlug ] }
+										</FormLegend>
+									</FormFieldset>
+									<div
+										onClick={ () => onChangeOption( JETPACK_SOCIAL_BASIC_PRODUCT?.productSlug ) }
+									>
+										<PaymentPlan
+											isMultiSiteIncompatible={ isMultiSiteIncompatible }
+											siteId={ siteId }
+											product={ JETPACK_SOCIAL_BASIC_PRODUCT as SelectorProduct }
+											showPlansOneBelowTheOther={ true }
+											isActive={
+												product?.productSlug === JETPACK_SOCIAL_BASIC_PRODUCT?.productSlug
+											}
+										/>
+									</div>
+									<p />
+									<div
+										onClick={ () => onChangeOption( JETPACK_SOCIAL_ADVANCED_PRODUCT?.productSlug ) }
+									>
+										<PaymentPlan
+											isMultiSiteIncompatible={ isMultiSiteIncompatible }
+											siteId={ siteId }
+											product={ JETPACK_SOCIAL_ADVANCED_PRODUCT as SelectorProduct }
+											showPlansOneBelowTheOther={ true }
+											isActive={
+												product?.productSlug === JETPACK_SOCIAL_ADVANCED_PRODUCT?.productSlug
+											}
+										/>
+									</div>
+								</div>
```
Can we extract this to a separate component and then conditionally render it here?

manzoorwanijk commented on client/my-sites/plans/jetpack-plans/product-lightbox/index.tsx:
```diff
@@ -187,11 +205,48 @@ const ProductLightbox: React.FC< Props > = ( {
+											{ PRODUCT_OPTIONS_HEADER[ product?.productSlug ] }
+										</FormLegend>
+									</FormFieldset>
+									<div
+										onClick={ () => onChangeOption( JETPACK_SOCIAL_BASIC_PRODUCT?.productSlug ) }
+									>
+										<PaymentPlan
+											isMultiSiteIncompatible={ isMultiSiteIncompatible }
+											siteId={ siteId }
+											product={ JETPACK_SOCIAL_BASIC_PRODUCT as SelectorProduct }
+											showPlansOneBelowTheOther={ true }
+											isActive={
+												product?.productSlug === JETPACK_SOCIAL_BASIC_PRODUCT?.productSlug
+											}
+										/>
+									</div>
```
For a11y and better keyboard navigation support, these should be real radio buttons that can be styled the way we want.

manzoorwanijk commented on client/my-sites/plans/jetpack-plans/product-store/items-list/all-items.tsx:
```diff
@@ -91,18 +92,30 @@ export const AllItems: React.FC< AllItemsProps > = ( {
+					const ctaHref =
+						item.productSlug === PRODUCT_JETPACK_SOCIAL_BASIC &&
+						config.isEnabled( 'jetpack-social/advanced-plan' )
+							? '#'
```
To make sure that the URL gets updated correctly and the page doesn't scroll to the top.

```suggestion
							? `#${ item.productSlug }`
```

manzoorwanijk commented on client/my-sites/plans/jetpack-plans/product-store/items-list/all-items.tsx:
```diff
@@ -91,18 +92,30 @@ export const AllItems: React.FC< AllItemsProps > = ( {
+						config.isEnabled( 'jetpack-social/advanced-plan' )
+							? '#'
+							: getCheckoutURL( item );
+					const onClickCtaUrl =
```
This is not a URL (string), it's rather a callback (function), thus `onClickCta` sounds a better name than `onClickCtaUrl`

manzoorwanijk reviewed (commented):
I see the a11y and keyboard navigation issue in payment plans for Social - https://github.com/Automattic/wp-calypso/pull/72051#discussion_r1073109650

spsiddarthan committed:
Move the product variant code to its own component.

spsiddarthan committed:
Fix type mismatch and get unit tests to pass.

spsiddarthan committed:
Fix method type.

manzoorwanijk reviewed (approved):
This looks good to me now.

spsiddarthan committed:
turn the advanced plan on in staging.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7754399

Thank you @spsiddarthan for including a screenshot in the description! This is really helpful for our translators.
